clean_code/pattern_algorithms.py

```diff
@@ -0,0 +1,124 @@
+
+from collections import Counter
+from math import factorial
+from bisect import insort
+
+"""This module supplies fast algorithms for pattern-related searches."""
+
+def find_n_biggest_numbers(numbers, amount=20):
+    if len(numbers) > amount:
+        biggest_numbers = sorted(numbers[:amount])
+    else:
+        return numbers
+
+
+    for number in numbers:
+        if number > biggest_numbers[0]:
+            insort(biggest_numbers, number)
+            del biggest_numbers[0]
+
+    return biggest_numbers
+
+
+def create_trimmed_string(base_string):
+    return ''.join(base_string[index] for index in range(1, len(base_string)) if base_string[index] != base_string[index-1])
+
+
+def find_longest_palindrome(base_string):
+    longest_pali_length = 0
+    longest_pali_start_index = 0
+    is_odd = True
+
+    for index in range(len(base_string) - 1):
+        current_length = _odd_palindrome_length(base_string, index)
+        if current_length > longest_pali_length:
+            longest_pali_length = current_length
+            longest_pali_start_index = index
+            is_odd = True
+
+        if current_length == 1 and base_string[index] == base_string[index+1]:
+            current_length = _even_palindrome_length(base_string, index, index+1)
+            if current_length > longest_pali_length:
+                longest_pali_length = current_length
+                longest_pali_start_index = index
+                is_odd = False
+    
+    begining = longest_pali_start_index - int(longest_pali_length/2)
+    end = longest_pali_start_index + int(longest_pali_length / 2) + 1
+
+    if is_odd:
+        return base_string[begining:end]
+    else:
+        return base_string[begining + 1:end]
+
+
+def _odd_palindrome_length(base_string, middle):
+    pali_length = 1
+    distance_from_middle = 1
+    while distance_from_middle + middle < len(base_string) and middle >= distance_from_middle:
+        if base_string[middle + distance_from_middle] == base_string[middle - distance_from_middle]:
+            pali_length += 2
+            distance_from_middle += 1
+        else:
+            return pali_length
+
+    return pali_length
+
+
+def _even_palindrome_length(base_string, middle_left, middle_right):
+    pali_length = 2
+    distance_from_middle = 1
+    while distance_from_middle + middle_right < len(base_string) and middle_left >= distance_from_middle:
+        if base_string[middle_right + distance_from_middle] == base_string[middle_left - distance_from_middle]:
+            pali_length += 2
+            distance_from_middle += 1
+        else:
+            return pali_length
+
+    return pali_length
+
+
+def generate_permutations_iterably(elements):
```

Comment:
This function has horrible efficiency (calculate permutations of string in length 8 takes 7 minutes!). Every permutation is generated over and over. I think you can generate only the needed permutations without repeating. But however, you can use  `_generate_new_permutations` only on the newly generated permutations and save a lot of time...